backend/database.js

```diff
@@ -52,57 +52,91 @@ export default ((sbp('sbp/selectors/register', {
     }
     // Number of entries pushed.
     let counter = 0
-    let { hash: currentHash, ...serverMeta } = (
-      await sbp('chelonia/db/getEntryMeta', contractID, height)
-    ) || {}
+    let currentHeight = height
+    let currentHash, serverMeta
     let prefix = ''
-    let ended = false
-    // NOTE: if this ever stops working you can also try Readable.from():
-    // https://nodejs.org/api/stream.html#stream_stream_readable_from_iterable_options
-    const stream = new Readable({
-      construct (cb) {
-        this.push('[')
-        cb()
-      },
-      read (): void {
-        if (ended) {
-          this.destroy()
-          return
+    const nextKeyOp = (() => {
+      let index
+
+      return async () => {
+        if (!index) {
+          index = (await sbp('chelonia.db/get', `_private_keyop_idx_${contractID}_${currentHeight - currentHeight % 10000}`))?.split('\x00')
         }
-        if (currentHash && counter < limit) {
-          sbp('chelonia/db/getEntry', currentHash).then(async entry => {
-            if (entry) {
-              const currentPrefix = prefix
-              prefix = ','
-              counter++
-              const { hash: newCurrentHash, ...newServerMeta } = (
-                await sbp('chelonia/db/getEntryMeta', contractID, entry.height() + 1)
-              ) || {}
-              this.push(
-                `${currentPrefix}"${strToB64(
-                  JSON.stringify({ serverMeta, message: entry.serialize() })
-                )}"`
-              )
-              currentHash = newCurrentHash
-              serverMeta = newServerMeta
-            } else {
-              this.push(']')
-              this.push(null)
-              ended = true
-            }
-          }).catch(e => {
-            console.error(`[backend] streamEntriesAfter: read(): ${e.message}:`, e.stack)
-            this.push(']')
-            this.push(null)
-            ended = true
-          })
+        const value = index?.find((h, i) => {
+          if (Number(h) >= currentHeight) {
+            // Remove values that no longer are relevant from the index
+            index = ((index: any): any[]).splice(i + 1)
+            return true
+          } else {
+            return false
+          }
+        })
+        if (value != null) {
+          const newHeight = Number(value)
+          currentHeight = newHeight
         } else {
-          this.push(']')
-          this.push(null)
-          ended = true
+          currentHeight = currentHeight - currentHeight % 10000 + 10000
+          index = undefined
+          if (currentHeight > latestHEADinfo.height) {
+            return false
+          } else {
+            return null
+          }
         }
+        return true
       }
-    })
+    })()
+    const fetchMeta = async () => {
```

Comment:
This returns different return values, presumably with differing meaning (which isn't documented), but the return values also don't seem to be used anywhere... So what's the point of returning a value that isn't ever used?

Reply:
If you look at previous iterations of `fetchMeta`, the return value was used, but then I refactored this to simplify the logic and a side effect is that the return value no longer needs to be used. I'd still leave it so that it's easier to start using it in the future if the need arises (but I can add a comment about what they signify).